raft.go

```diff
@@ -977,41 +975,49 @@ func (r *Raft) runLeader() {
 	r.leaderLoop()
 }
 
-// startReplication is a helper to setup state and start async replication to a peer.
-// It's safe to call this with the local server (in that case, it does nothing.)
-func (r *Raft) startReplication(server Server) {
-	if server.ID == r.localID {
-		return
-	}
+// startStopReplication will set up state and start asynchronous replication to
+// new peers, and stop replication to removed peers. Before removing a peer,
+// it'll instruct the replication routines to try to replicate to the current
+// index.
+func (r *Raft) startStopReplication() {
+	inConfig := make(map[ServerID]bool, len(r.leaderState.replState))
 	lastIdx := r.getLastIndex()
-	s := &followerReplication{
-		peer:        server,
-		commitment:  r.leaderState.commitment,
-		stopCh:      make(chan uint64, 1),
-		triggerCh:   make(chan struct{}, 1),
-		currentTerm: r.getCurrentTerm(),
-		nextIndex:   lastIdx + 1,
-		lastContact: time.Now(),
-		notifyCh:    make(chan struct{}, 1),
-		stepDown:    r.leaderState.stepDown,
-	}
-	r.leaderState.replState[server.ID] = s
-	r.goFunc(func() { r.replicate(s) })
-	asyncNotifyCh(s.triggerCh)
-}
 
-// stopReplication is a helper to stop the replication goroutine for the given
-// peer, after it's tried to replicate up through the given index.
-// It's safe to call this with the local server (in that case, it does nothing.)
-func (r *Raft) stopReplication(server Server, index uint64) {
-	if server.ID == r.localID {
-		return
+	// Start replication goroutines that need starting
+	for _, server := range r.configurations.latest.Servers {
+		if server.ID == r.localID {
+			continue
+		}
+		inConfig[server.ID] = true
+		r.logger.Printf("[INFO] raft: Added peer %v, starting replication", server)
+		if _, present := r.leaderState.replState[server.ID]; !present {
```

Comment:
I was initially worried that this wouldn't create a new replication task if there was one in effect that was draining (if you added and removed a peer quickly), because the old code didn't have the `!present` check, but we delete from `replState` whenever we stop replication so I think we are good.